base_layer/core/src/base_node/states/fetching_horizon_state.rs

```diff
@@ -49,60 +81,185 @@ impl HorizonInfo {
             target: LOG_TARGET,
             "Synchronising kernel merkle mountain range to pruning horizon."
         );
-        if let Err(e) = self.synchronize_kernel_mmr().await {
+        if let Err(e) = self.synchronize_kernel_mmr(shared).await {
             return StateEvent::FatalError(format!("Synchronizing kernel MMR failed. {}", e));
         }
 
         info!(target: LOG_TARGET, "Synchronising range proof MMR to pruning horizon.");
-        if let Err(e) = self.synchronize_range_proof_mmr().await {
+        if let Err(e) = self.synchronize_range_proof_mmr(shared).await {
             return StateEvent::FatalError(format!("Synchronizing range proof MMR failed. {}", e));
         }
 
         info!(target: LOG_TARGET, "Synchronising TXO MMR to pruning horizon.");
-        if let Err(e) = self.synchronize_output_mmr().await {
+        if let Err(e) = self.synchronize_output_mmr(shared).await {
             return StateEvent::FatalError(format!("Synchronizing output MMR failed. {}", e));
         }
 
         info!(target: LOG_TARGET, "Synchronising headers to pruning horizon.");
-        if let Err(e) = self.synchronize_headers().await {
+        if let Err(e) = self.synchronize_headers(shared).await {
             return StateEvent::FatalError(format!("Synchronizing block headers failed. {}", e));
         }
 
         info!(target: LOG_TARGET, "Synchronising kernels to pruning horizon.");
-        if let Err(e) = self.synchronize_kernels().await {
+        if let Err(e) = self.synchronize_kernels(shared).await {
             return StateEvent::FatalError(format!("Synchronizing kernels failed. {}", e));
         }
 
         info!(target: LOG_TARGET, "Synchronising UTXO set at pruning horizon.");
-        if let Err(e) = self.synchronize_utxo_set().await {
+        if let Err(e) = self.synchronize_utxo_set(shared).await {
             return StateEvent::FatalError(format!("Synchronizing UTXO set failed. {}", e));
         }
 
         debug!(target: LOG_TARGET, "Pruning horizon state has synchronised");
         StateEvent::HorizonStateFetched
     }
 
-    async fn synchronize_headers(&mut self) -> Result<(), String> {
-        Err("unimplemented".into())
+    // Retrieve the full base mmr state for the specified MmrTree by performing multiple queries and reassembling the
+    // full state from the received chunks.
+    async fn download_mmr_base_state<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+        tree: MmrTree,
+    ) -> Result<MutableMmrLeafNodes, String>
+    {
+        let leaf_nodes_sync_chunk_size = shared.config.horizon_sync_config.leaf_nodes_sync_chunk_size as u64;
+        let mut index = 0;
+        let mut complete_leaf_count = leaf_nodes_sync_chunk_size;
+        let mut base_state = MutableMmrLeafNodes::new(Vec::new(), Bitmap::create());
+        while index < complete_leaf_count {
```

Comment:
I tlooks like It'll only ever run one iteration of this loop because

complete_leaf_count <- leaf_nodes_sync_chunk_size
and 
index += leaf_nodes_sync_chunk_size

Reply:
With the first query of the fetch_mmr_state in the while loop, the complete_leaf_count will be updated to the true received count, resulting in the loop running until all the entries have been obtained.

Reply:
Oh, see it. Thanks